src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -212,26 +212,39 @@ InflightData GetInflightData (NSUrlSessionTask task)
 				var inflight = default (InflightData);
 
 				lock (sessionHandler.inflightRequestsLock)
-					if (sessionHandler.inflightRequests.TryGetValue (task, out inflight))
+					if (sessionHandler.inflightRequests.TryGetValue (task, out inflight)) {
+						// ensure that we did not cancel the request, if we did, do cancel the task
+						if (inflight.CancellationToken.IsCancellationRequested)
```

Comment:
You mean `inflight?.CancellationToken` ? I don't see why, `inflight` can't be null -  a new instance is added with a task, so if the task is found then you have it (it's under a lock too).

As for `CancellationToken?.IsCancellationRequested` then it's not needed, CancellationToken being a struct.